stub_uploader/get_version.py

```
@@ -83,28 +60,90 @@ def check_exists(distribution: str) -> bool:
+    # version will satisfy the version_spec (defined precisely by the `compatible`
+    # specifier below). This allows users to have expectations of what a stub package
+    # will contain based on the upstream version they're targeting.
+    if version_spec.endswith(".*"):
+        compatible = SpecifierSet(f"=={version_spec}")
+    else:
+        compatible = SpecifierSet(f"=={version_spec}.*")
+
+    # See what the base version and specificity we used in METADATA.toml
+    version_base = Version(version_spec.removesuffix(".*"))
+    specificity = len(version_base.release)
+
+    if max_published.epoch > 0 or version_base.epoch > 0:
+        raise NotImplementedError("Epochs in versions are not supported")
+
+    # We'll try to bump the fourth part of the release. So e.g. if our version_spec is
+    # 1.1, we'll release 1.1.0.0, 1.1.0.1, 1.1.0.2, ...
+    # But if our version_spec is 5.6.7.8, we'll release 5.6.7.8.0, 5.6.7.8.1, ...
+    increment_specificity = max(specificity + 1, 4)
+
+    if version_base.release < max_published.release[:specificity]:
+        # Our published versions have gone too far ahead the upstream version
+        # So we can't guarantee our second property.
+        # In practice, this will only happen if the specificity of version_spec is
+        # changed or we change our versioning scheme.
+        # For example, version_base=1.2, max_published=1.3.0.4, return 1.3.0.5
+        increment_specificity = max(increment_specificity, len(max_published.release))
+        incremented = list(max_published.release)
+        extend_to_specificity(incremented, increment_specificity)
+        incremented[-1] += 1
+
+        incremented_version = Version(".".join(map(str, incremented)))
+        assert incremented_version > max_published
+        # But can't keep versioning compatible with upstream...
+        assert incremented_version not in compatible
+        return incremented_version
```

I know that this block is basically the point of the PR, *but* could we just raise an exception here for now and wait one or two weeks before adding it in a second PR? I just want to make sure that the other changes break anything before making the functional changes. (This also makes this PR a pure refactoring.)

---

This still needs addressing.